superset/db_engine_specs/base.py

```diff
@@ -472,7 +472,10 @@ def get_engine(
         schema: Optional[str] = None,
         source: Optional[utils.QuerySource] = None,
     ) -> Engine:
-        return database.get_sqla_engine(schema=schema, source=source)
+        with database.get_sqla_engine_with_context(
+            schema=schema, source=source
+        ) as engine:
+            return engine
```

Comment:
This won't work; before returning the context manager will be exited before the functions returns. It's equivalent to:

```python
with database.get_sqla_engine_with_context(...) as engine:
    a = engine
return a
```

Reply:
so should i just `return database._ get_sqla_engine()`

Reply:
> so should i just `return database._ get_sqla_engine()`

If you do that, then the SSH tunneling wouldn't be setup in places that call `DBEngineSpec.get_engine()`, defeating the purpose of the changes in this PR. You need to convert this into a context manager as well (and update everything downstream), so that the setup/teardown works as expected.

You can play with a simple Python script to understand how this works:

```python
from contextlib import contextmanager

@contextmanager
def get_sqla_engine_with_context():
    print("enabling ssh tunnel")
    yield 42
    print("disabling ssh tunnel")

def get_engine():
    with get_sqla_engine_with_context() as engine:
        return engine

print("start")
engine = get_engine()
print("I have the engine, I can now work on it")
print("end")
```

Running the code above prints:

```before
start
enabling ssh tunnel
disabling ssh tunnel
I have the engine, I can now work on it
end
```